### PR TITLE
Fix #26 NPE during generateToString

### LINKED ISSUE #26: NPE during compilation
I get an error during compilation

```
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: .../MyModel.kt
The root cause java.lang.NullPointerException was thrown at: net.afanasev.sekret.kotlin.SekretClassBuilder.generateToString(SekretClassBuilder.kt:107)
```

I'm confused because I'm not using the `@Secret` annotation on this class - maybe it's because of the inline value class, `ModelId`?

```kotlin
data class MyModel(
    override val id: ModelId,
    override val name: String,
    override val description: String?,
) : Product


interface Product {
    val id: ModelId?
    val name: String?
    val description: String?
}


@JvmInline
value class ModelId(val value: String) : CharSequence by value
```

* Kotlin 1.7.0
* Sekret 0.1.2

```none
> Task :my-application:compileKotlin FAILED
e: org.jetbrains.kotlin.backend.common.BackendException: Backend Internal error: Exception during IR lowering
File being compiled: .../MyModel.kt
The root cause java.lang.NullPointerException was thrown at: net.afanasev.sekret.kotlin.SekretClassBuilder.generateToString(SekretClassBuilder.kt:107)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException(CodegenUtil.kt:239)
	at org.jetbrains.kotlin.backend.common.CodegenUtil.reportBackendException$default(CodegenUtil.kt:235)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:68)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invoke(performByIrFile.kt:55)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invoke(performByIrFile.kt:41)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.CompositePhase.invoke(PhaseBuilders.kt:29)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.CompilerPhaseKt.invokeToplevel(CompilerPhase.kt:43)
	at org.jetbrains.kotlin.backend.jvm.JvmIrCodegenFactory.invokeCodegen(JvmIrCodegenFactory.kt:284)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.runCodegen(KotlinToJVMBytecodeCompiler.kt:355)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli(KotlinToJVMBytecodeCompiler.kt:136)
	at org.jetbrains.kotlin.cli.jvm.compiler.KotlinToJVMBytecodeCompiler.compileModules$cli$default(KotlinToJVMBytecodeCompiler.kt:60)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:157)
	at org.jetbrains.kotlin.cli.jvm.K2JVMCompiler.doExecute(K2JVMCompiler.kt:52)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:94)
	at org.jetbrains.kotlin.cli.common.CLICompiler.execImpl(CLICompiler.kt:43)
	at org.jetbrains.kotlin.cli.common.CLITool.exec(CLITool.kt:101)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:477)
	at org.jetbrains.kotlin.incremental.IncrementalJvmCompilerRunner.runCompiler(IncrementalJvmCompilerRunner.kt:127)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally(IncrementalCompilerRunner.kt:366)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileIncrementally$default(IncrementalCompilerRunner.kt:311)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl$rebuild(IncrementalCompilerRunner.kt:111)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compileImpl(IncrementalCompilerRunner.kt:165)
	at org.jetbrains.kotlin.incremental.IncrementalCompilerRunner.compile(IncrementalCompilerRunner.kt:74)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.execIncrementalCompiler(CompileServiceImpl.kt:625)
	at org.jetbrains.kotlin.daemon.CompileServiceImplBase.access$execIncrementalCompiler(CompileServiceImpl.kt:101)
	at org.jetbrains.kotlin.daemon.CompileServiceImpl.compile(CompileServiceImpl.kt:1739)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:359)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
	at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:562)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:796)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:677)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:676)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.NullPointerException
	at net.afanasev.sekret.kotlin.SekretClassBuilder.generateToString(SekretClassBuilder.kt:107)
	at net.afanasev.sekret.kotlin.SekretClassBuilder.done(SekretClassBuilder.kt:89)
	at org.jetbrains.kotlin.backend.jvm.codegen.ClassCodegen.generate(ClassCodegen.kt:195)
	at org.jetbrains.kotlin.backend.jvm.FileCodegen.lower(JvmPhases.kt:44)
	at org.jetbrains.kotlin.backend.common.phaser.FileLoweringPhaseAdapter.invoke(PhaseBuilders.kt:120)
	at org.jetbrains.kotlin.backend.common.phaser.FileLoweringPhaseAdapter.invoke(PhaseBuilders.kt:116)
	at org.jetbrains.kotlin.backend.common.phaser.NamedCompilerPhase.invoke(CompilerPhase.kt:96)
	at org.jetbrains.kotlin.backend.common.phaser.PerformByIrFilePhase.invokeSequential(performByIrFile.kt:65)
	... 42 more
```

### PR DESCRIPTION
**Pull request by aSemy**
Fixes #26

Depends on #27 #28 

I'm not sure what the problem was - but this seems to fix it when I test locally.

I wasn't able to re-create the issue so there's no test :(

If you want to preview this PR, then I've published it to the forked repo - just add a Maven repo to both the regular and plugin repositories

```kotlin
// settings.gradle.kts
fun RepositoryHandler.sekret() {
    // Access a PR version of Sekret while awaiting official fix: https://github.com/aafanasev/sekret/issues/26
    maven("https://raw.githubusercontent.com/aSemy/sekret/main/.m2") {
        content { includeGroup("net.afanasev") }
    }
}

@Suppress("UnstableApiUsage") // centralised repositories are incubating
dependencyResolutionManagement {

    repositories {
        mavenCentral()
        sekret()
    }

    pluginManagement {
        repositories {
            gradlePluginPortal()
            mavenCentral()
            sekret()
        }
    }
}
```

### CODE REVIEW

File: kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt [hunk 1]
```diff
@@ -102,23 +111,23 @@ class SekretClassBuilder(
...
-            val name = entry.key
-            val fieldInfo = entry.value!!
-            val desc = fieldInfo.first
-            val needToHide = fieldInfo.second
-            val isNullable = fieldInfo.third
+        fields.onEachIndexed { index, (name, fieldInfo) ->
+
+            if (fieldInfo == null) {
+                println("WARNING $index/$name has no field info")
+                return@onEachIndexed
```
Comment (aafanasev) [inline kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:118]: This will skip `value class` fields in toString output

```MyModel(, name=name, description=null)```

whereas it's expected to be

```MyModel(id=ModelId(value=1), name=name, description=null)```

https://pl.kotl.in/N376re2nv

Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:118]: Ah, thanks for catching this! 

I've created some failing test cases based on the example you provided. Interestingly the fields are only skipped if the data class implements an interface.

File: kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt [hunk 2]
```diff
@@ -74,8 +78,13 @@ class SekretClassBuilder(
                     descriptor: String,
                     isInterface: Boolean
                 ) {
-                    if (opcode == Opcodes.INVOKEVIRTUAL && owner == origin.classDescriptor() && name.startsWith("get")) {
-                        fields[name.substring(3).replaceFirstChar { it.lowercase() }] = null
+                    if (opcode == Opcodes.INVOKEVIRTUAL
+                        && owner == origin.classDescriptor()
+                        && name.startsWith("get")
+                    ) {
+                        val keyName = name.substring(3).replaceFirstChar { it.lowercase() }
```
Comment (aafanasev) [inline kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: Got it. In case of value classes, the compiler still adds the wrapped value to the class but renames its getter to `get<field name>-<something random>` so it can't be accessed:

```
  // access flags 0x12
  private final Ljava/lang/CharSequence; id
  @Lorg/jetbrains/annotations/NotNull;() // invisible

  // access flags 0x1
  public getId-vQR88SI()Ljava/lang/CharSequence;
  @Lorg/jetbrains/annotations/NotNull;() // invisible
   L0
    LINENUMBER 6 L0
    ALOAD 0
    GETFIELD net/afanasev/sekret/sample/MyModel.id : Ljava/lang/CharSequence;
    ARETURN
   L1
    LOCALVARIABLE this Lnet/afanasev/sekret/sample/MyModel; L0 L1 0
    MAXSTACK = 1
    MAXLOCALS = 1
```

That's why the toString generator cannot find any info about `id` field as it was added as `id-vS88I`.

Given the dash symbol (`-`) is not allowed in property/function names, I think it's safe to remove the tail from the getter like `name.substring(3).substringBefore("-").replaceFirstChar { it.lowercase() }`

WDYT?

Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: Ah yes, here's the docs on the mangling https://kotlinlang.org/docs/inline-classes.html#mangling

I'll make the change and see if it works!

Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: I can't seem to get any change I make to be picked up by the tests... I think I might have broken something in the Gradle config update? Possibly [here](https://github.com/aafanasev/sekret/pull/27/files#diff-3a54cc5f752d8af679e43524c289aba5f924edf6606d747c1e9b03e09c8744daR24-R31)? But I still see`kotlin-plugin-0.2.2.jar` is generated correctly.

Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: Also, I wonder if `JvmDeclarationOriginKind.UNBOX_METHOD_OF_INLINE_CLASS` or `JvmDeclarationOriginKind.ERASED_INLINE_CLASS` might be more reliable than checking the method name for a dash? But I don't know enough about how compiler plugins work to figure it out.

Comment (aafanasev) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: > Also, I wonder if `JvmDeclarationOriginKind.UNBOX_METHOD_OF_INLINE_CLASS` or `JvmDeclarationOriginKind.ERASED_INLINE_CLASS` might be more reliable than checking the method name for a dash? But I don't know enough about how compiler plugins work to figure it out.

Yeah, if you're given this info it'd be more reliable

Comment (aafanasev) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: > I can't seem to get any change I make to be picked up by the tests... I think I might have broken something in the Gradle config update? Possibly [here](https://github.com/aafanasev/sekret/pull/27/files#diff-3a54cc5f752d8af679e43524c289aba5f924edf6606d747c1e9b03e09c8744daR24-R31)? But I still see`kotlin-plugin-0.2.2.jar` is generated correctly.

Can you commit your changes? Will try to run tests on CI.

Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: I didn't make any real changes, I was just commenting out code and throwing exceptions, to see if my changes would be picked up

Comment (aafanasev) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: Tweaked the class builder in https://github.com/aafanasev/sekret/commit/8627ef0056f7fcc286bf56e4df7fe11ee2b27315 but still failing.

`./gradlew clean :sample:test`

```kotlin
// should print non-annotated properties
expect: <MySubModel(id=ModelId(value=111), name=Name(value=name), description=Description(value=description), extraField=extra-field)> 
actual: <MySubModel(id=111, name=name, description=description, extraField=extra-field)>
```

Indeed, it should print out wrapper classes as well https://pl.kotl.in/bLRmg48nu. Not sure if we have access to value classes. Will look closely in weekends.

Comment (aafanasev) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:85]: @aSemy, https://github.com/aafanasev/sekret/commit/244947f4604b379e1ac237bf3f5732f4b2b4a57d adds value classes and fixes all tests, have a look. Would you like to apply those changes in this PR?

**Comment by aSemy** [comment]
Works like a dream! I added a few more tests cases for edge-cases - they worked first time!

File: kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt [hunk 3]
```diff
@@ -28,6 +26,8 @@ class SekretClassBuilder(
     private val fields = linkedMapOf<String, FieldInfo?>()
     private var generateToString = false
 
+    private var error = ""
```
Comment (aSemy) [inline kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:29]: I think this can be removed?

Comment (aafanasev) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt:29]: Yes, can be removed. Used just for debugging.

Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt]: done!

File: kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt (hunk 2 above)
Comment (aSemy) [reply kotlin-plugin/src/main/kotlin/net/afanasev/sekret/kotlin/SekretClassBuilder.kt]: done!

**Review by aafanasev — APPROVED** [review]
